v2/bundle/x509bundle/bundle.go

```diff
@@ -1,65 +1,176 @@
+
 // Bundle is a collection of trusted public key material for a trust domain.
 type Bundle struct {
+	trustDomain spiffeid.TrustDomain
+
+	rootsMtx sync.RWMutex
+	roots    []*x509.Certificate
 }
 
 // New creates a new bundle
 func New(trustDomain spiffeid.TrustDomain) *Bundle {
-	panic("not implemented")
+	return &Bundle{
+		trustDomain: trustDomain,
+		roots:       []*x509.Certificate{},
+	}
 }
 
 // Load loads a Bundle from a file on disk.
 func Load(trustDomain spiffeid.TrustDomain, path string) (*Bundle, error) {
-	panic("not implemented")
+	fileBytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, x509bundleErr.New("unable to load X.509 bundle file: %w", err)
+	}
+
+	return Parse(trustDomain, fileBytes)
 }
 
 // Read decodes a bundle from a reader.
 func Read(trustDomain spiffeid.TrustDomain, r io.Reader) (*Bundle, error) {
-	panic("not implemented")
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, x509bundleErr.New("unable to read X.509 bundle: %v", err)
+	}
+
+	return Parse(trustDomain, b)
 }
 
 // Parse parses a bundle from bytes.
 func Parse(trustDomain spiffeid.TrustDomain, b []byte) (*Bundle, error) {
-	panic("not implemented")
+	bundle := New(trustDomain)
+	for {
+		if len(b) == 0 {
+			break
+		}
+		pemBlock, pemBytes := pem.Decode(b)
+		b = pemBytes
+		if pemBlock == nil {
+			return nil, x509bundleErr.New("no PEM data found while decoding block")
+		}
+
+		if pemBlock.Type != certType {
+			return nil, x509bundleErr.New("block does not contain %q type, current type is: %q", certType, pemBlock.Type)
+		}
+
+		cert, err := x509.ParseCertificate(pemBlock.Bytes)
+		if err != nil {
+			return nil, x509bundleErr.New("cannot parse certificate: %v", err)
+		}
+		bundle.AddX509Root(cert)
+	}
+	return bundle, nil
 }
 
 // TrustDomain returns the trust domain of the bundle.
 func (b *Bundle) TrustDomain() spiffeid.TrustDomain {
-	panic("not implemented")
+	return b.trustDomain
 }
 
 // X509Roots returns the X.509 roots in the bundle.
 func (b *Bundle) X509Roots() []*x509.Certificate {
-	panic("not implemented")
+	b.rootsMtx.RLock()
+	defer b.rootsMtx.RUnlock()
+	return b.roots
 }
 
 // AddX509Root adds an X.509 root to the bundle. If the root already
 // exists in the bundle, the contents of the bundle will remain unchanged.
-func (b *Bundle) AddX509Root(*x509.Certificate) {
-	panic("not implemented")
+func (b *Bundle) AddX509Root(root *x509.Certificate) {
+	b.rootsMtx.Lock()
+	defer b.rootsMtx.Unlock()
+
+	for _, r := range b.roots {
+		if areCertsEqual(r, root) {
+			return
+		}
+	}
+
+	b.roots = append(b.roots, root)
 }
 
 // RemoveX509Root removes an X.509 root to the bundle.
-func (b *Bundle) RemoveX509Root(*x509.Certificate) {
-	panic("not implemented")
+func (b *Bundle) RemoveX509Root(root *x509.Certificate) {
+	b.rootsMtx.Lock()
+	defer b.rootsMtx.Unlock()
+
+	for i, r := range b.roots {
+		if areCertsEqual(r, root) {
+			//remove element from slice (slice order is not preserved)
```

Comment:
I've looked over this a few times and keep rationalizing how not preserving order should be ok, but every time it gives me pause. Considering removing a root should be an infrequent, likely non-bulk operation, how would you feel doing a less performant "remove" just to preserve order? Somebody might be surprised by the change in ordering if they are diff'ing serialized versions.

Reply:
That is a good point. Removing elements in that way will cause more headaches than benefits. I'll change it